src/libraries/System.Private.CoreLib/src/System/Threading/LowLevelSpinWaiter.cs

```diff
@@ -5,8 +5,52 @@
 
 namespace System.Threading
 {
+    /// <summary>
+    /// A lightweight spin-waiter intended to be used as the first-level wait for a condition before the user forces the thread
+    /// into a wait state, and where the condition to be checked in each iteration is relatively cheap, like just an interlocked
+    /// operation.
+    ///
+    /// Used by the wait subsystem on Unix, so this class cannot have any dependencies on the wait subsystem.
+    /// </summary>
     internal struct LowLevelSpinWaiter
     {
+        private int _spinningThreadCount;
+
+        public bool SpinWaitForCondition(Func<bool> condition, int spinCount, int sleep0Threshold)
+        {
+            Debug.Assert(condition != null);
+
+            int processorCount = Environment.ProcessorCount;
+            int spinningThreadCount = Interlocked.Increment(ref _spinningThreadCount);
+            try
+            {
+                // Limit the maximum spinning thread count to the processor count to prevent unnecessary context switching
+                // caused by an excessive number of threads spin waiting, perhaps even slowing down the thread holding the
+                // resource being waited upon
+                if (spinningThreadCount <= processorCount)
+                {
+                    // For uniprocessor systems, start at the yield threshold since the pause instructions used for waiting
```

Comment:
The benefit of spin-waiting would also be lower on containers affinitized to 1 processor, although the yield may allow other processes to do work on the core, if the spin-waiting is waiting for a change that would be made by a different thread in the same process (often the case) then it would not be able to see the change until it allows a different thread to run. `Sleep` may be slightly cheaper than the actual wait sometimes (used in `LowLevelLock` in this case, where the actual wait involves acquiring a mutex and waiting on a condition variable), but it may be slightly worse in other cases where there is no other thread to schedule (unlikely in this case). In several other cases spin-waiting is avoided on single-processor machines. The spin-waiting done here is minimal and probably decent for now.